plugin-azure-server/src/main/kotlin/jetbrains/buildServer/clouds/azure/arm/KubePodNameGenerator.kt

```diff
@@ -0,0 +1,134 @@
+import jetbrains.buildServer.serverSide.executors.ExecutorServices
+import jetbrains.buildServer.util.EventDispatcher
+import jetbrains.buildServer.util.FileUtil
+import jetbrains.buildServer.util.StringUtil
+import org.jetbrains.annotations.NotNull
+import java.io.File
+import java.io.IOException
+import java.nio.charset.Charset
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentMap
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReadWriteLock
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+
+class KubePodNameGenerator(@NotNull eventDispatcher: EventDispatcher<BuildServerListener>,
+                           @NotNull  executorServices: ExecutorServices) {
+
+    private val myCounters: ConcurrentMap<String, AtomicInteger> = ConcurrentHashMap()
+    private val myIndexTouchedMaps: ConcurrentMap<String, AtomicBoolean> = ConcurrentHashMap()
+    private var myIndexStorage: File = File(ServerPaths.PLUGIN_DATA_DIRECTORY_NAME, "azureKubeIndex")
+    private val myIsAvailable: AtomicBoolean = AtomicBoolean(false)
+    private val myLock: ReadWriteLock = ReentrantReadWriteLock()
+
+    init{
+        if (!myIndexStorage.exists()) {
+            myIndexStorage.mkdirs()
+        }
+        if (!myIndexStorage.isDirectory) {
+            throw CloudException("Unable to create a directory for azure arm plugin to monitor VM indexes")
+        }
+        myIsAvailable.set(true)
+
+        loadIndexes(myIndexStorage)
+
+        eventDispatcher.addListener(object : BuildServerAdapter() {
+            override fun serverShutdown() {
+                myIsAvailable.set(false)
+                storeIndexes()
+            }
+        })
+        executorServices.normalExecutorService.scheduleWithFixedDelay(this::storeIndexes, 60, 60, TimeUnit.SECONDS)
+    }
+
+    @Synchronized
+    private fun loadIndexes(@NotNull indexStorage: File) {
+        val indexes = indexStorage.listFiles() ?: return
+        for (indexFile in indexes) {
+            if (!indexFile.name.endsWith(".idx")) continue
+            val indexName = indexFile.name.substring(0, indexFile.name.length - 4)
+            try {
+                val indexValueFromFile = FileUtil.readText(indexFile).toInt()
+                if (indexValueFromFile > 0) {
+                    myCounters.putIfAbsent(indexName, AtomicInteger(0))
+                    myCounters[indexName]?.set(indexValueFromFile)
+                }
+            } catch (e: IOException) {
+                continue
+            }
+        }
+    }
+
+    private fun storeIndexes() {
+        // wait for generation operations to finish:
+        try {
+            if (!myLock.writeLock().tryLock(100, TimeUnit.MILLISECONDS)) {
+                Loggers.AGENT.warn("Waited more than 100ms to store Kube indexes")
+            }
```

Comment:
These three lines probably ought to be moved outside of the `try finally` otherwise in the event gaining the write lock fails, a call to `unlock()` will still occur resulting in a `IllegalMonitorStateException`

Reply:
locks can through an InterruptedException, they need to be inside the try block.

Reply:
Oh actually in this case any time the `tryLock` fails a `IllegalMonitorStateException` will be thrown. Although after all the work of this plugin has completed, so maybe it would be ok.

The java doc for ReentrantLock shows the lock being held outside:
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html
This makes sense since in the event the thread is interrupted while waiting to get the lock, an exception will be raised and a call to `unlock` will not go ahead.

I wrote two tests that demonstrate the exception raised from the lock being acquired within the `try` https://gist.github.com/LukeButters/aa82af50657e2c479c6956625dc58637